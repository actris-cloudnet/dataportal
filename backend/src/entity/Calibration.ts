import { Column, Entity, BeforeInsert, BeforeUpdate } from "typeorm";

@Entity()
export class Calibration {
  @Column({ primary: true })
  instrumentPid!: string;

  @Column({ type: "date", primary: true })
  measurementDate!: string;

  @Column({ type: "jsonb" })
  data!: any;

  @Column()
  createdAt!: Date;

  @Column()
  updatedAt!: Date;

  @BeforeInsert()
  updateDateCreation() {
    this.createdAt = new Date();
    this.updatedAt = this.createdAt;
  }

  @BeforeUpdate()
  updateDateUpdate() {
    this.updatedAt = new Date();
  }
}
