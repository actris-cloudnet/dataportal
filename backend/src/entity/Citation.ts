import {Column, Entity, ManyToMany, ManyToOne, PrimaryColumn, PrimaryGeneratedColumn} from 'typeorm'
import {Site} from './Site'
import {Model} from './Model'


@Entity()
export class Citation {

  @PrimaryColumn()
  id!: string

  @Column({type: 'text'})
  acknowledgements!: string

}

@Entity()
export class RegularCitation extends Citation {

  @ManyToMany(_ => Site)
  sites!: Site[]
}

@Entity()
export class ModelCitation extends Citation {

  @ManyToMany(_ => Model)
  models!: Model[]
}
