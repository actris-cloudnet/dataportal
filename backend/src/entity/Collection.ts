import {
  BeforeInsert,
  BeforeUpdate,
  Column,
  Entity,
  JoinTable,
  ManyToMany,
  PrimaryColumn,
} from 'typeorm'
import {File} from './File'
import { v4 as generateUuidV4 } from 'uuid'

@Entity()
export class Collection {

  @PrimaryColumn('uuid')
  uuid!: string

  @ManyToMany(_type => File)
  @JoinTable()
  files!: File[]

  @Column({default: ''})
  title!: string

  @Column({default: ''})
  pid!: string

  @Column({default: 0})
  downloadCount!: number

  @Column()
  createdAt!: Date

  @Column()
  updatedAt!: Date

  @BeforeInsert()
  setCreatedAt() {
    this.createdAt = new Date()
    this.updatedAt = this.createdAt
  }

  @BeforeUpdate()
  setUpdatedAt() {
    this.updatedAt = new Date()
  }

  constructor(files: File[]) {
    this.uuid = generateUuidV4()
    this.files = files
  }
}
