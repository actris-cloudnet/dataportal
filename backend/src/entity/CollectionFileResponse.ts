import {File, ModelFile, RegularFile} from './File'

export class CollectionFileResponse {

  uuid: string
  measurementDate: Date
  site: string
  siteId: string
  product: string
  productId: string
  modelId: string | null
  size: number
  volatile: boolean
  legacy: boolean

  constructor(file: RegularFile | ModelFile) {
    this.uuid = file.uuid
    this.measurementDate = file.measurementDate
    this.site = file.site.humanReadableName
    this.siteId = file.site.id
    this.product = file.product.humanReadableName
    this.productId = file.product.id
    this.modelId = 'model' in file ? file.model.id : null
    this.size = file.size
    this.volatile = file.volatile
    this.legacy = file.legacy
  }
}
