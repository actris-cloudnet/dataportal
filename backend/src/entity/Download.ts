import { Column, Entity, PrimaryGeneratedColumn, ViewEntity, ViewColumn } from "typeorm";

export enum ObjectType {
  Product = "product",
  Collection = "collection",
  Raw = "raw",
}

@Entity()
export class Download {
  @PrimaryGeneratedColumn()
  id!: number;

  @Column()
  objectType!: ObjectType;

  @Column("uuid")
  objectUuid!: string;

  @Column()
  ip!: string;

  @Column({ type: "char", length: 2, nullable: true })
  country!: string | null;

  @Column()
  createdAt!: Date;

  constructor(objectType: ObjectType, objectUuid: string, ip: string, country?: string, createdAt = new Date()) {
    this.objectType = objectType;
    this.objectUuid = objectUuid;
    this.ip = ip;
    this.country = country || null;
    this.createdAt = createdAt;
  }
}

@ViewEntity({
  expression: `
    SELECT
      "createdAt"::date AS "downloadDate",
      "ip",
      "country",
      "measurementDate",
      "productType",
      "siteId",
      SUM("downloads") AS "downloads"
    FROM download
    JOIN (
      SELECT uuid, "measurementDate", "productType", "siteId", COUNT(*) AS "downloads"
      FROM (
        SELECT uuid, "measurementDate", 'observation' AS "productType", "siteId"
        FROM regular_file
        JOIN product_variable USING ("productId")
        WHERE product_variable."actrisName" IS NOT NULL
        UNION ALL
        SELECT uuid, "measurementDate", 'model' AS "productType", "siteId"
        FROM model_file
        JOIN product_variable USING ("productId")
        WHERE product_variable."actrisName" IS NOT NULL
      ) AS file
      GROUP BY uuid, "measurementDate", "productType", "siteId"
      UNION ALL
      SELECT "collectionUuid" AS uuid, "measurementDate", "productType", "siteId", COUNT(*) AS "downloads"
      FROM (
        SELECT "collectionUuid", "measurementDate", 'observation' AS "productType", "siteId"
        FROM collection_regular_files_regular_file
        JOIN regular_file ON "regularFileUuid" = regular_file.uuid
        JOIN product_variable USING ("productId")
        WHERE product_variable."actrisName" IS NOT NULL
        UNION ALL
        SELECT "collectionUuid", "measurementDate", 'model' AS "productType", "siteId"
        FROM collection_model_files_model_file
        JOIN model_file ON "modelFileUuid" = model_file.uuid
        JOIN product_variable USING ("productId")
        WHERE product_variable."actrisName" IS NOT NULL
      ) AS collection_file
      GROUP BY "collectionUuid", "measurementDate", "productType", "siteId"
    ) object ON "objectUuid" = object.uuid
    WHERE
      ip NOT IN ('', '::ffff:127.0.0.1')
      AND ip NOT LIKE '192.168.%'
      AND ip NOT LIKE '193.166.223.%'
    GROUP BY "downloadDate", "ip", "country", "measurementDate", "productType", "siteId"
  `,
  materialized: true,
})
export class DownloadStats {
  @ViewColumn()
  downloadDate!: Date;

  @ViewColumn()
  measurementDate!: Date;

  @ViewColumn()
  productType!: string;

  @ViewColumn()
  siteId!: string;

  @ViewColumn()
  downloads!: number;
}
