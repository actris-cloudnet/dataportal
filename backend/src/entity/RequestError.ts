export interface RequestError {
  status: number
  errors: Array<string> | string
}
