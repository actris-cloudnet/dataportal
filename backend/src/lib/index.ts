import { FindOperator, Repository } from "typeorm";
import { basename } from "path";
import { NextFunction, Request, Response } from "express";
import { ModelFile, RegularFile } from "../entity/File";
import { File } from "../entity/File";
import { SearchFileResponse } from "../entity/SearchFileResponse";
import { SearchFile } from "../entity/SearchFile";
import { Upload } from "../entity/Upload";
import axios from "axios";
import { SiteType } from "../entity/Site";
import env from "./env";
import ReadableStream = NodeJS.ReadableStream;
import maxmind, { CountryResponse, OpenOpts, Reader } from "maxmind";
import { randomBytes } from "crypto";
import { Collection } from "../entity/Collection";
import { InstrumentInfo } from "../entity/Instrument";

export const stringify = (obj: any): string => JSON.stringify(obj, null, 2);

const DATETIME_FORMAT =
  /^(?<year>\d\d\d\d)-(?<month>\d\d)-(?<day>\d\d)(T(?<hours>\d\d):(?<minutes>\d\d):(?<seconds>\d\d)(\.(?<fraction>\d+))?(Z|\+00:00)?)?$/;

export function isValidDate(obj: any): boolean {
  if (obj instanceof Date) {
    return !isNaN(obj.getTime());
  }
  if (typeof obj !== "string") {
    return false;
  }
  const match = DATETIME_FORMAT.exec(obj);
  if (!match || !match.groups) {
    return false;
  }
  const year = parseInt(match.groups.year);
  const month = parseInt(match.groups.month) - 1;
  const day = parseInt(match.groups.day);
  const hours = match.groups.hours ? parseInt(match.groups.hours) : 0;
  const minutes = match.groups.minutes ? parseInt(match.groups.minutes) : 0;
  const seconds = match.groups.seconds ? parseInt(match.groups.seconds) : 0;
  const milliseconds = match.groups.fraction ? parseInt(match.groups.fraction.slice(0, 3).padEnd(3, "0")) : 0;
  const date = new Date(year, month, day, hours, minutes, seconds, milliseconds);
  return (
    date.getFullYear() == year &&
    date.getMonth() == month &&
    date.getDate() == day &&
    date.getHours() == hours &&
    date.getMinutes() == minutes &&
    date.getSeconds() == seconds &&
    date.getMilliseconds() == milliseconds
  );
}

export const tomorrow = () => {
  const tomorrow = new Date();
  tomorrow.setDate(tomorrow.getDate() + 1);
  return tomorrow;
};

export const toArray = <T>(obj: T | T[] | undefined): T[] | null => {
  if (!obj) return null;
  else if (!Array.isArray(obj)) return [obj];
  return obj;
};

export const hideTestDataFromNormalUsers = (dbQuery: any, req: Request) =>
  req.query.developer !== undefined ? dbQuery : dbQuery.andWhere("not :type = ANY(site.type)", { type: SiteType.TEST });

export const convertToSearchResponse = (file: SearchFile) => new SearchFileResponse(file);

export const convertToReducedResponse = (parameters: (keyof SearchFileResponse)[]) => (file: SearchFile) =>
  parameters.reduce((acc, cur) => ({ ...acc, [cur]: convertToSearchResponse(file)[cur] }), {});

export const augmentFile = (includeS3path: boolean) => (file: RegularFile | ModelFile) => ({
  ...file,
  downloadUrl: `${env.DP_BACKEND_URL}/download/${getDownloadPathForFile(file)}`,
  filename: basename(file.s3key),
  s3path: includeS3path ? getS3pathForFile(file) : undefined, // TODO: not used anymore?
  model: "model" in file ? file.model : undefined,
  software: parseSoftware(file),
  timeliness: calcTimeliness(file),
  sourceRegularFiles: undefined,
  sourceModelFiles: undefined,
  sourceFileIds: getSourceFileIds(file),
});

const parseSoftware = (file: RegularFile | ModelFile) =>
  file.software?.map((software) => ({
    id: software.code,
    version: software.version,
    title: `${software.humanReadableName || software.code} ${software.version}`,
    url: software.url,
  }));

function calcTimeliness(file: RegularFile | ModelFile) {
  if (!file.stopTime) {
    const daysBetween = daysBetweenDates(file.updatedAt, file.measurementDate);
    return daysBetween <= 3 ? "nrt" : "scheduled";
  }
  const msInHour = 1000 * 60 * 60;
  const msInDay = msInHour * 24;
  const diff = file.updatedAt.getTime() - file.stopTime.getTime();
  if (diff < msInHour * 3) return "rrt";
  else if (diff < msInDay * 3) return "nrt";
  else return "scheduled";
}

function getSourceFileIds(file: RegularFile | ModelFile) {
  if (file instanceof ModelFile || !file.sourceRegularFiles || !file.sourceModelFiles) {
    return undefined;
  }
  return [...file.sourceRegularFiles.map((file) => file.uuid), ...file.sourceModelFiles.map((file) => file.uuid)];
}

export const ssAuthString = () => "Basic " + Buffer.from(`${env.DP_SS_USER}:${env.DP_SS_PASSWORD}`).toString("base64");

export const getBucketForFile = (file: File) => (file.volatile ? "cloudnet-product-volatile" : "cloudnet-product");

export const uploadBucket = "cloudnet-upload";

export const generateS3keyForUpload = (upload: Upload) => `${upload.site.id}/${upload.uuid}/${upload.filename}`;

export const getS3pathForUpload = (upload: Upload) => `/${uploadBucket}/${upload.s3key}`;

export const getS3pathForFile = (file: File) => `/${getBucketForFile(file)}/${file.s3key}`;

export const getS3pathForImage = (s3key: string) => `/cloudnet-img/${s3key}`;

export const getDownloadPathForFile = (file: File) => `product/${file.uuid}/${file.filename}`;

export async function checkFileExists(s3path: string) {
  const headers = {
    Authorization: ssAuthString(),
  };
  return axios.head(`${env.DP_SS_URL}${s3path}`, { headers });
}

const translateKeyVal = (key: string, val: string | number | boolean | Date, acc: any, prefix: string) => {
  if (key == "file_siteId" || key == "file_productId") return {}; // Ignore unneeded fields
  const regexp = new RegExp(`^${prefix}_`);
  key = key.replace(regexp, "");
  val = val instanceof Date && key == "measurementDate" ? val.toISOString().split("T")[0] : val;
  let subKey;
  [key, ...subKey] = key.split("_");
  subKey = subKey.join("_");
  if (!subKey) return { [key]: val };
  else
    return {
      [key]: {
        ...acc[key],
        ...{ [subKey]: val },
      },
    };
};

export const transformRawFile = (obj: any, prefix: string): RegularFile | ModelFile | SearchFile => {
  return Object.keys(obj).reduce(
    (acc: Record<string, any>, key) => ({
      ...acc,
      ...translateKeyVal(key, obj[key], acc, prefix),
    }),
    {},
  ) as RegularFile | ModelFile | SearchFile;
};

export const dateforsize = async (
  repo: Repository<any>,
  table: string,
  req: Request,
  res: Response,
  _: NextFunction,
) => {
  const query = req.query as any;
  const startDate = new Date(query.startDate);
  const sizeBytes = parseInt(query.targetSize) * 1024 * 1024 * 1024;

  const result = await repo.query(
    `SELECT "updatedAt" FROM (
    SELECT "updatedAt", sum(size) OVER (ORDER BY "updatedAt")
    FROM ${table} where "updatedAt" > $1) as asd
  WHERE sum > $2 LIMIT 1`,
    [startDate, sizeBytes],
  );

  if (result.length == 0) return res.sendStatus(400);
  return res.send(result[0].updatedAt);
};

export function streamHandler(stream: ReadableStream, res: Response, prefix: string, augmenter?: (f: any) => any) {
  res.header("content-type", "application/json");
  res.write("[");
  let objectSent = false;
  stream.on("data", (data) => {
    if (objectSent) res.write(",");
    else objectSent = true;
    const transformedFile = transformRawFile(data, prefix);
    const augmentedFile = augmenter ? augmenter(transformedFile) : transformedFile;
    res.write(JSON.stringify(augmentedFile));
  });
  stream.on("end", () => {
    res.write("]");
    res.end();
  });
}

export function getIpLookup(options?: OpenOpts): Promise<Reader<CountryResponse>> {
  return maxmind.open<CountryResponse>(env.GEOLITE2_COUNTRY_PATH, options);
}

export function randomString(length: number): string {
  return randomBytes(Math.ceil(length / 2))
    .toString("hex")
    .slice(0, length);
}

export function validateInstrumentPid(input: any): string {
  if (typeof input !== "string") {
    return "must be string";
  }
  if (!input.startsWith("https://")) {
    return "must be HTTPS";
  }
  if (!input.startsWith("https://hdl.handle.net/")) {
    return "must be Handle";
  }
  return "";
}

export function getFileLandingPage(file: RegularFile | ModelFile): string {
  return `${env.DP_FRONTEND_URL}/file/${file.uuid}`;
}

export function getCollectionLandingPage(collection: Collection): string {
  return `${env.DP_FRONTEND_URL}/collection/${collection.uuid}`;
}

export function getObjectLandingPage(object: RegularFile | ModelFile | Collection): string {
  return object instanceof Collection ? getCollectionLandingPage(object) : getFileLandingPage(object);
}

export function truncateList(list: string[], limit: number, placeholder: string) {
  if (list.length < limit + 2) {
    return list;
  }
  return [...list.slice(0, limit), `${list.length - limit} other ${placeholder}`];
}

export function formatList(parts: string[], conjunction: string): string {
  if (parts.length <= 2) {
    return parts.join(conjunction);
  }
  return parts.slice(0, -1).join(", ") + conjunction + parts[parts.length - 1];
}

// TODO: Workaround for TypeORM 0.3
export function ArrayEqual<T>(value: T[] | FindOperator<T>): FindOperator<any> {
  return new FindOperator("equal", value as any);
}

export function daysBetweenDates(a: Date, b: Date): number {
  a = new Date(a);
  b = new Date(b);
  a.setHours(0, 0, 0, 0);
  b.setHours(0, 0, 0, 0);
  return Math.floor(Math.abs(a.getTime() - b.getTime()) / (24 * 60 * 60 * 1000));
}

export function fixInstrument(instrument: string, instrumentInfo: InstrumentInfo) {
  // "chm15k" and "chm15kx" are mixed up so consider them equal
  if (instrument == "chm15k" || instrument == "chm15kx") {
    if (instrumentInfo.instrument.id == "chm15k" || instrumentInfo.instrument.id == "chm15kx") {
      return instrument;
    } else {
      return null;
    }
  }
  // Convert "mira" to "mira-35"
  if (instrument == "mira") {
    instrument = "mira-35";
  }
  if (instrument != instrumentInfo.instrument.id) {
    return null;
  }
  return instrument;
}

export function isStringArray(value: any): value is string[] {
  return Array.isArray(value) && value.every((item) => typeof item === "string");
}
