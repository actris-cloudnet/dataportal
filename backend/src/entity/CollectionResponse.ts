import {Collection} from './Collection'
import {SearchFileResponse} from './SearchFileResponse'
import {convertToCollectionFileResponse, convertToSearchResponse} from '../lib'
import {File, ModelFile, RegularFile} from './File'
import {CollectionFileResponse} from './CollectionFileResponse'

export class CollectionResponse {

  uuid: string
  files: CollectionFileResponse[]
  title: string
  pid: string
  downloadCount: number
  createdAt: Date
  updatedAt: Date

  constructor(coll: Collection) {
    const sortedFiles = (coll.regularFiles as unknown as (RegularFile | ModelFile)[]).concat(coll.modelFiles)
      .sort((a, b) => a.measurementDate > b.measurementDate ? 1 : -1)
    this.uuid = coll.uuid
    this.files = sortedFiles.map(convertToCollectionFileResponse)
    this.title = coll.title
    this.pid = coll.pid
    this.downloadCount = coll.downloadCount
    this.createdAt = coll.createdAt
    this.updatedAt = coll.updatedAt
  }
}
