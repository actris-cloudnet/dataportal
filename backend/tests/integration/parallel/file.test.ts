import { backendPublicUrl } from "../../lib";
import axios from "axios";
import { RequestError } from "../../../src/entity/RequestError";
import { describe, expect, it } from "@jest/globals";

const testUuid = "9e04d8ef-0f2b-4823-835d-33e458403c67";

describe("/api/files/:uuid", () => {
  const url = `${backendPublicUrl}files/`;
  const expectedBody404: RequestError = {
    status: 404,
    errors: ["No files match this UUID"],
  };

  it("request succeeds on instrument file", async () => {
    const res = await axios.get(`${url}38092c00-161d-4ca2-a29d-628cf8e960f6`);
    expect(res.data).toMatchSnapshot();
  });

  it("request succeeds on model file", async () => {
    const res = await axios.get(`${url}b5d1d5af-3667-41bc-b952-e684f627d91c`);
    expect(res.data).toMatchSnapshot();
  });

  it("responds with a 404 on test file if in normal mode", async () => {
    return expect(axios.get(url + testUuid)).rejects.toMatchObject({ response: { data: expectedBody404 } });
  });

  it("responds with a 400 if invalid uuid", async () => {
    const expectedBody = { ...expectedBody404, ...{ errors: ["Not found: invalid UUID"] } };
    return expect(axios.get(`${url}kisuli`)).rejects.toMatchObject({ response: { data: expectedBody } });
  });

  it("request succeeds on a test file in developer mode", async () => {
    return expect(axios.get(url + testUuid, { params: { developer: "" } })).resolves.toBeTruthy();
  });
});
