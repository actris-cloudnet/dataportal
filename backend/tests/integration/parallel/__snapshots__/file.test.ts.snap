// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/api/files/:uuid request succeeds on instrument file 1`] = `
{
  "checksum": "298688b011a511f8f0e9353371cf73ee86f60c89b0e02c6931d8c05542c64cdb",
  "createdAt": "2020-02-20T10:56:19.382Z",
  "downloadUrl": "http://localhost:3000/api/download/product/38092c00-161d-4ca2-a29d-628cf8e960f6/20181115_mace-head_mira.nc",
  "dvasId": null,
  "dvasUpdatedAt": null,
  "errorLevel": null,
  "filename": "20181115_mace-head_mira.nc",
  "format": "HDF5 (NetCDF4)",
  "instrument": {
    "allowedTags": [],
    "humanReadableName": "METEK MIRA-35 cloud radar",
    "id": "mira-35",
    "shortName": "MIRA",
    "type": "radar",
  },
  "instrumentInfo": {
    "model": "MIRA",
    "name": "Bucharest MIRA",
    "owners": [
      "University of Bucharest",
    ],
    "pid": "https://hdl.handle.net/123/bucharest-mira",
    "serialNumber": null,
    "type": "cloud radar",
    "uuid": "0b3a7fa0-4812-4964-af23-1162e8b3a665",
  },
  "instrumentPid": "https://hdl.handle.net/123/bucharest-mira",
  "legacy": false,
  "measurementDate": "2018-11-15",
  "pid": "",
  "product": {
    "experimental": false,
    "humanReadableName": "Radar",
    "id": "radar",
    "type": [
      "instrument",
    ],
  },
  "s3key": "38092c00-161d-4ca2-a29d-628cf8e960f6/20181115_mace-head_mira.nc",
  "site": {
    "actrisId": null,
    "altitude": 16,
    "country": "Ireland",
    "countryCode": "IE",
    "countrySubdivisionCode": null,
    "description": null,
    "dvasId": "mchd",
    "gaw": "MHD",
    "humanReadableName": "Mace Head",
    "id": "mace-head",
    "latitude": 53.326,
    "longitude": -9.9,
    "stationName": null,
    "type": [
      "cloudnet",
    ],
  },
  "size": "12200657",
  "software": [
    {
      "id": "cloudnetpy",
      "title": "CloudnetPy 1.0.4",
      "url": null,
      "version": "1.0.4",
    },
  ],
  "sourceFileIds": [],
  "startTime": "2018-11-15T10:56:19.382Z",
  "stopTime": "2018-11-15T11:56:19.382Z",
  "timeliness": "scheduled",
  "tombstoneReason": null,
  "updatedAt": "2020-02-20T10:56:19.382Z",
  "uuid": "38092c00-161d-4ca2-a29d-628cf8e960f6",
  "version": "123",
  "volatile": true,
}
`;

exports[`/api/files/:uuid request succeeds on model file 1`] = `
{
  "checksum": "055bcd9deae26851992c3da6352844fb9443203c48ae4f4ad8f1aa50ef2ab26f",
  "createdAt": "2020-02-20T10:52:59.073Z",
  "downloadUrl": "http://localhost:3000/api/download/product/b5d1d5af-3667-41bc-b952-e684f627d91c/20141205_mace-head_ecmwf.nc",
  "dvasId": null,
  "dvasUpdatedAt": null,
  "errorLevel": null,
  "filename": "20141205_mace-head_ecmwf.nc",
  "format": "NetCDF3",
  "legacy": false,
  "measurementDate": "2020-12-05",
  "model": {
    "forecastEnd": null,
    "forecastStart": null,
    "humanReadableName": "ECMWF forecast",
    "id": "ecmwf",
    "optimumOrder": 0,
    "sourceModelId": null,
  },
  "pid": "",
  "product": {
    "experimental": false,
    "humanReadableName": "Model",
    "id": "model",
    "type": [
      "model",
    ],
  },
  "s3key": "b5d1d5af-3667-41bc-b952-e684f627d91c/20141205_mace-head_ecmwf.nc",
  "site": {
    "actrisId": 1337,
    "altitude": 93,
    "country": "Romania",
    "countryCode": "RO",
    "countrySubdivisionCode": null,
    "description": null,
    "dvasId": "INO",
    "gaw": null,
    "humanReadableName": "Bucharest",
    "id": "bucharest",
    "latitude": 44.348,
    "longitude": 26.029,
    "stationName": null,
    "type": [
      "cloudnet",
    ],
  },
  "size": "500452",
  "software": [],
  "startTime": null,
  "stopTime": null,
  "timeliness": "scheduled",
  "tombstoneReason": null,
  "updatedAt": "2020-02-20T10:52:59.073Z",
  "uuid": "b5d1d5af-3667-41bc-b952-e684f627d91c",
  "version": "123",
  "volatile": true,
}
`;

exports[`/api/files/:uuid returns list of file versions 1`] = `
[
  {
    "createdAt": "2020-02-20T10:59:58.449Z",
    "uuid": "8bb32746-faf0-4057-9076-ed2e698dcf36",
  },
  {
    "createdAt": "2020-02-20T10:49:58.449Z",
    "uuid": "6cb32746-faf0-4057-9076-ed2e698dcf36",
  },
  {
    "createdAt": "2020-02-20T10:39:58.449Z",
    "uuid": "22b32746-faf0-4057-9076-ed2e698dcc34",
  },
  {
    "createdAt": "2020-02-19T10:59:58.449Z",
    "uuid": "1bb32746-faf0-4057-9076-ed2e698dcf36",
  },
  {
    "createdAt": "2020-02-19T10:59:58.449Z",
    "uuid": "2bb32746-faf0-4057-9076-ed2e698dcf36",
  },
]
`;

exports[`/api/files/:uuid returns list of file versions with PID and checksum 1`] = `
[
  {
    "checksum": "64d86b26f642a0befdcd7088ff38caab00ea3d95933f4ccc430179b4a29b74d3",
    "createdAt": "2020-02-20T10:59:58.449Z",
    "pid": "",
    "uuid": "8bb32746-faf0-4057-9076-ed2e698dcf36",
  },
  {
    "checksum": "43d86b26f642a0befdcd7088ff38caab00ea3d95933f4ccc430179b4a29b74d3",
    "createdAt": "2020-02-20T10:49:58.449Z",
    "pid": "",
    "uuid": "6cb32746-faf0-4057-9076-ed2e698dcf36",
  },
  {
    "checksum": "22d86b26f642a0befdcd7088ff38caab00ea3d95933f4ccc430179b4a29b74d3",
    "createdAt": "2020-02-20T10:39:58.449Z",
    "pid": "http://pidservice.example/pidhandle",
    "uuid": "22b32746-faf0-4057-9076-ed2e698dcc34",
  },
  {
    "checksum": "14d86b26f642a0befdcd7088ff38caab00ea3d95933f4ccc430179b4a29b74d3",
    "createdAt": "2020-02-19T10:59:58.449Z",
    "pid": "",
    "uuid": "1bb32746-faf0-4057-9076-ed2e698dcf36",
  },
  {
    "checksum": "24d86b26f642a0befdcd7088ff38caab00ea3d95933f4ccc430179b4a29b74d3",
    "createdAt": "2020-02-19T10:59:58.449Z",
    "pid": "",
    "uuid": "2bb32746-faf0-4057-9076-ed2e698dcf36",
  },
]
`;
